flang/lib/Frontend/FrontendActions.cpp

```diff
@@ -535,6 +535,34 @@ void EmitObjAction::ExecuteAction() {
   llvmModule_->setDataLayout(TM->createDataLayout());
   assert(TM && "Failed to create TargetMachine");
 
+  // If the output stream is a file, generate it and define the corresponding
+  // output stream. If a pre-defined output stream is available, we will use
+  // that instead.
+  //
+  // NOTE: `os` is a smart pointer that will be destroyed at
+  // the end of this method. However, it won't be written to until
+  // `CodeGenPasses` is destroyed. The corresponding raw pointer might be later
+  // wrapped into a smart pointer inside `LLVMTargetMachine::createMCStreamer`
+  // (this depends on whether an object or an assembly file is generated). By
+  // defining `os` before `CodeGenPasses` we make sure that the output
+  // stream won't be destroyed before it is written to. This only applies when
+  // an output file is used (i.e. there is no pre-defined output stream).
+  // TODO: Revisit once the new PM is ready (i.e. when `CodeGenPasses` is
+  // updated to us it).
+  std::unique_ptr<llvm::raw_pwrite_stream> os;
+  if (ci.IsOutputStreamNull()) {
+      // Get the output buffer/file
+      std::string fileExt = (_act == BackendAct::Backend_EmitAssembly) ? "s" : "o";
+      os = ci.CreateDefaultOutputFile(
+      /*Binary=*/true, /*InFile=*/GetCurrentFileOrBufferName(), fileExt);
```

Comment:
Do you want the output file format to be Binary even when outputting assembly?

Reply:
Good catch, thanks! WIll fix this shortly!